Fix #8048 : make `docker images repository:tag` work

vdemeester opened the pull request:
Related to #8048, Make command like "docker images ubuntu:14.04" work and filter out the image with the given tag. The idea was to be able to do the following :

``` bash
$ docker images java
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
java                8                   308e519aac60        6 days ago          824.5 MB
java                7                   493d82594c15        3 months ago        656.3 MB
java                latest              2711b1d6f3aa        5 months ago        603.9 MB
$ docker images java:8
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
java                8                   308e519aac60        6 days ago          824.5 MB
$ docker images java:7
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
java                7                   493d82594c15        3 months ago        656.3 MB
$ docker images java:9 # o
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
```

Added two integration tests in `docker_cli_images_test.go`.

Signed-off-by: Vincent Demeester vincent@sbr.pm

runcom commented:
+1 for usability

vdemeester commented:
@runcom what do you mean ? :smile:

runcom commented:
I mean you  can actually do `docker images reponame` but the tag filter is done in `--filter` option. Pretty much like the label filter is done.

vdemeester commented:
Ah right.. Using a ``--filter`` or ``--tag`` would be possible (and probably have a cleaner implementation). Now from a user point of vue, as we display tag as ``image:tag`` and as we use it, for example when pulling (``docker pull image:tag``), it feels natural/straighfoward to do the same with ``ps`` — my 2 cents :smiley:

runcom commented:
@vdemeester yes that makes sense (no `--tag` tho, I deleted from my previous comment), but maybe we should have had the image filter in filters too, just wondering eh..

vdemeester commented:
@runcom yep I kinda agree — but I think it's *a little too late* to make this kind of change as it would probably break a lot of stuff (in user's script & co).

runcom commented:
Yes changing that was out of the discussion of course (I was just thinking out loud)

jfrazelle commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -25,6 +25,55 @@ func (s *DockerSuite) TestImagesEnsureImageIsListed(c *check.C) {
 }
 
+func (s *DockerSuite) TestImagesEnsureImageWithTagIsListed(c *check.C) {
+	_, err := buildImage("imagewithtag:v1",
```
this needs a testRequires(c, Network)

vdemeester commented:
@jfrazelle updated :wink:

icecrime commented:
Code LGTM

icecrime commented:
~~LGTM~~

Sorry, I'm missing something:

```
root@aa78ad12d8d9:/go/src/github.com/docker/docker# docker images 
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
busybox             latest              8c2e06607696        11 weeks ago        2.43 MB
root@aa78ad12d8d9:/go/src/github.com/docker/docker# docker images :latest 
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
```

Why is that?

icecrime commented:
LGTM

icecrime commented on graph/list.go:
```diff
@@ -70,13 +70,29 @@ func (s *TagStore) Images(config *ImagesConfig) ([]*types.Image, error) {
 		}
 		for ref, id := range repository {
 			imgRef := utils.ImageReference(repoName, ref)
+			if !strings.Contains(imgRef, filterTagName) {
```
I'm surprised that this doesn't return a match for `busybox:latest`:

```
root@aa78ad12d8d9:/go/src/github.com/docker/docker# docker images busybox:la    
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
```

icecrime replied:
Oh, that's because of the `continue` on line 87, nevermind...

vdemeester replied:
oh I didn't add a tests on that (trying with a part of an existing tag), might be worth to add (would document a bit more the behavior).

crosbymichael commented:
@vdemeester isn't it much simpler to just do `docker images | grep latest`

vdemeester commented:
@crosbymichael It is (at least for me) and that's what I'm using (like ``docker images java | grep latest``). 

But :
* For the given example, ``docker images java | grep 8`` can give me back more than one image, or even the wrong on (in the example below, if I don't have ``java:8``, it will still get me one result) :

```bash
$ docker images java | grep 8                                                                                                  
java                8                   308e519aac60        6 weeks ago         824.5 MB
java                7                   493d82594c15        5 months ago        656.3 MB
```

* I tend to see a lot of people trying to do some ``docker images myimage:tag`` and be surprised it doesn't work — it feels a little inconsistent.

Of course, there is other ways to do that too, using ``inspect`` (see https://github.com/docker/docker/issues/8048#issuecomment-76514389) or else.

tiborvass commented:
@vdemeester sorry needs another rebase :S

vdemeester commented:
@tiborvass done :stuck_out_tongue_closed_eyes:

cpuguy83 commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -18,6 +18,44 @@ func (s *DockerSuite) TestImagesEnsureImageIsListed(c *check.C) {
 }
 
+func (s *DockerSuite) TestImagesEnsureImageWithTagIsListed(c *check.C) {
+	testRequires(c, Network)
```
How come this needs networking?

vdemeester replied:
@cpuguy83 you're completely right, I don't remember why I put this and also think it's not needed… I will remove it :wink:.

vdemeester replied:
Ah I remember now : https://github.com/docker/docker/pull/13501#discussion_r31480008, @jfrazelle told me :stuck_out_tongue_closed_eyes:

cpuguy83 commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -18,6 +18,44 @@ func (s *DockerSuite) TestImagesEnsureImageIsListed(c *check.C) {
+	_, err := buildImage("imagewithtag:v1",
+		`FROM scratch
+		MAINTAINER dockerio1`, true)
+	if err != nil {
```
c.Assert(err, check.IsNil)

cpuguy83 commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -18,6 +18,44 @@ func (s *DockerSuite) TestImagesEnsureImageIsListed(c *check.C) {
+	_, err = buildImage("imagewithtag:v2",
+		`FROM scratch
+		MAINTAINER dockerio1`, true)
+	if err != nil {
```
c.Assert(err, check.IsNil)

tianon commented:
For when `grep` isn't quite enough, I usually turn to `awk`:
```console
$ docker images java | awk -F '  +' '$2 == "8" { print }'
java                8                   5282faca75e8        18 hours ago        817.5 MB
$ docker images java | awk -F '  +' '$2 == "8" { print $3 }'
5282faca75e8
```

(I admit, not terribly obvious and I'm personally +1 for the idea behind this PR, but figured I'd throw it out there.)

cpuguy83 commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -18,6 +18,40 @@ func (s *DockerSuite) TestImagesEnsureImageIsListed(c *check.C) {
+		MAINTAINER dockerio1`, true)
+	c.Assert(err, check.IsNil)
+
+	time.Sleep(time.Second)
```
Sorry, also what's up with this sleep?

vdemeester replied:
I did the same as the other tests (like ``TestImagesOrderedByCreationDate`` for example) that were doing that. I'm not quit sure about the reason.. It seems it was already there when these tests were in ``integration``. 
@cpuguy83 I could test without it if it works well and then *maybe* remove it for the other tests (in another PR then) if we think it's not needed at all :sweat_smile:.

cpuguy83 replied:
In another PR is fine.
1sec is a lot of time, especially if a lot of tests are doing this sleep.
Thanks for clarifying!

cpuguy83 commented:
LGTM

thaJeztah commented:
Does this need documentation changes? (Not checked yet, just a quick note some we don't forget :relaxed:)

calavera commented:
code LGTM, but this is a new public behavior and it needs documentation.

Moving to docs-review, please do NOT merge without documenting it.

vdemeester committed:
Fix #8048 : make `docker images repository:tag` work

Make command like "docker images ubuntu:14.04" work and filter out the
image with the given tag.

Closes #8048.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
Added some docs (and in the ``man`` page too :wink:). 
/ping @moxiegirl @thaJeztah :blush:

thaJeztah commented on docs/reference/commandline/images.md:
```diff
@@ -52,6 +52,35 @@ uses up the `VIRTUAL SIZE` listed only once.
 
+### Listing images by name and tag
+
+The ``docker images`` command takes an optional ``[REPOSITORY[:TAG]]``
```
Just a note: wondering why you're using multiple back-ticks; in general we're using single back-ticks for this (but this should probably work as well) :D

vdemeester replied:
@thaJeztah I have no idea :sweat_smile:. Long time ago, I probably used a markdown lib. where using simple back-ticks didn't work and I got used to it.. I didn't know it was working with simple back-tick `test` :sweat:… (the lazy box inside of me is pretty happy, one keystroke less to type :smile_cat:)

thaJeztah commented on docs/reference/commandline/images.md:
```diff
@@ -52,6 +52,35 @@ uses up the `VIRTUAL SIZE` listed only once.
+
+> NOTE: the lookup on ``REPOSITORY`` is done on the entire word. This means
+> ``docker images jav`` won't match the image ``java``.
+
```
My attempt - forgive me if @moxiegirl comes with a better suggestion :relaxed: 

```Markdown
The `docker images` command takes an optional `[REPOSITORY[:TAG]]` argument
that allows you to filter the list of images for the given repository and tag.

If a `REPOSITORY` is specified, but no `TAG`, the `docker images` command
will list all images in the given repository.

For example, to list all images in the "java" repository, run this command :

    $ docker images java
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
    java                8                   308e519aac60        6 days ago          824.5 MB
    java                7                   493d82594c15        3 months ago        656.3 MB
    java                latest              2711b1d6f3aa        5 months ago        603.9 MB


If both `REPOSITORY` and `TAG` are provided, only images matching that
repository and tag will be listed. If nothing matches `REPOSITORY[:TAG]`,
the list will be empty.

To find all local images in the "java" repository with tag "8" you can use:

    $ docker images java:8
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
    java                8                   308e519aac60        6 days ago          824.5 MB

    $ docker images java:0 # There is no such image with this tag
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE

Filtering on `REPOSITORY` performs an "exact match". This means that, for example,
`docker images jav` will not match the image `java`.

```

vdemeester replied:
It's already better than mine so I'll take it :+1: :stuck_out_tongue_closed_eyes:

thaJeztah replied:
:tada:

thaJeztah commented:
docs LGTM (but I'm biased), so

ping @moxiegirl for review :+1:

moxiegirl commented on docs/reference/commandline/images.md:
```diff
@@ -52,6 +52,39 @@ uses up the `VIRTUAL SIZE` listed only once.
     postgres                  latest              746b819f315e        4 days ago          213.4 MB
 
+### Listing images by name and tag
+
```
@vdemeester nicely done contribution! For the docs, I want to remove the "filtering" approach as this command does have actual filter feature.  Also, the exact match behavior should be explained earlier. Finally, there is no future in technical writing --- :-D or rarely.  

--->  https://gist.github.com/moxiegirl/13d3928a828389b73f74

### Listing images by name and tag

The `docker images` command takes an optional `[REPOSITORY[:TAG]]` argument
that restricts the list to images that match the argument. If you specify 
`REPOSITORY`but no `TAG`, the `docker images` command lists all images in the
given repository.

For example, to list all images in the "java" repository, run this command :

    $ docker images java
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
    java                8                   308e519aac60        6 days ago          824.5 MB
    java                7                   493d82594c15        3 months ago        656.3 MB
    java                latest              2711b1d6f3aa        5 months ago        603.9 MB

The `[REPOSITORY[:TAG]]` value must be an "exact match". This means that, for example,
`docker images jav` does not match the image `java`.

	If both `REPOSITORY` and `TAG` are provided, only images matching that
repository and tag are listed.  To find all local images in the "java"
repository with tag "8" you can use:

    $ docker images java:8
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
    java                8                   308e519aac60        6 days ago          824.5 MB

If nothing matches `REPOSITORY[:TAG]`, the list is empty.
	
    $ docker images java:0 
    REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE

thaJeztah replied:
> Finally, there is no future in technical writing --- :-D or rarely.

It took me a moment to interpret that correctly. It sounded a bit "end of an era" 
😄

vdemeester replied:
@moxiegirl thanks :+1: ! I tend to use *future* more than I should too (even in my native language so :sweat:)
@thaJeztah it took me a moment to interpret it correctly too :stuck_out_tongue_closed_eyes:.

moxiegirl replied:
The punster rides again!  :zap:

thaJeztah commented on man/docker-images.1.md:
```diff
@@ -61,6 +61,11 @@ The list will contain the image repository name, a tag for the image, and an
 
+The ``docker images`` command takes an optional ``[REPOSITORY[:TAG]]``
+argument that will filter the list of images for the given repository.
+
```
Oh, darn. Almost forgot the man page; can you update this description as well, based on the new text?

vdemeester replied:
@thaJeztah yes sir ! :stuck_out_tongue_closed_eyes:

thaJeztah replied:
@vdemeester merci monsieur demeester :fr:

thaJeztah commented:
re-LGTM

ping @moxiegirl for the merge

moxiegirl commented:
@thaJeztah My policy is never to merge a PR with code.  :facepunch: could happen.  @tianon or @tiborvass

thaJeztah commented:
@moxiegirl you didn't give your LGTM yet, that's why I didn't merge :wink:

tianon commented:
LGTM

moxiegirl commented:
@thaJeztah I thought it at you? Didn't you get it? :crystal_ball:

thaJeztah commented:
Pushing the button :-)

tonistiigi commented:
@vdemeester docs mention "exact match" but we are using `path.Match` to compare to repo name what allows different kind of wildcards. Could we just compare the string values instead?

vdemeester commented:
@tonistiigi hum probably yes, I could try that :wink:
